File: js_tools/read_file.js
return function(args) {
  if (typeof args.path !== "string") throw new Error("INVALID_ARGUMENT: Missing mandatory field: path");
  const path = args.path;
  if (path.includes("..") || path.startsWith("/")) {
    throw new Error("SECURITY_VIOLATION: Path traversal (..) or absolute paths are not allowed.");
  }

  function parseOptionalInteger(value, fieldName) {
    if (value === undefined || value === null) return null;
    if (typeof value === "number" && Number.isInteger(value)) return value;
    if (typeof value === "string" && /^-?\d+$/.test(value.trim())) {
      return Number.parseInt(value, 10);
    }
    throw new Error("INVALID_ARGUMENT: " + fieldName + " must be an integer");
  }

  const parsed_start_line = parseOptionalInteger(args.start_line, "start_line");
  const parsed_end_line = parseOptionalInteger(args.end_line, "end_line");

  let cmd = "cat " + shell_escape(path);
  const res = __os_run(cmd);
  if (res.exit_code !== 0) {
    throw new Error("Could not open file: " + res.stderr);
  }

  let lines = res.stdout.split("\n");
  if (res.stdout.endsWith("\n")) lines.pop();

  const start_line = parsed_start_line === null ? 1 : parsed_start_line;
  const end_line = parsed_end_line === null ? lines.length : parsed_end_line;

  if (start_line > end_line) {
    throw new Error("INVALID_ARGUMENT: start_line (" + start_line + ") cannot be greater than end_line (" + end_line + ")");
  }

  if (start_line > lines.length) {
    return "";
  }

  let result_lines = lines.slice(Math.max(0, start_line - 1), end_line);
  if (args.line_numbers) {
    result_lines = result_lines.map((line, i) => (Math.max(1, start_line) + i) + ": " + line);
  }

  let body = result_lines.join("\n");
  if (body.length > 0) body += "\n";
  return body;
};

File: core/tools/read_file.cpp
#include "core/tool_executor.h"

#include <fstream>
#include <vector>

#include "absl/status/status.h"
#include "absl/strings/match.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_join.h"

#include "core/status_macros.h"
#include "core/tools/common.h"
#include "core/json_utils.h"

namespace slop {
absl::StatusOr<std::string> ToolExecutor::HandleReadFile(const nlohmann::json& args) {
  auto path = json_get<std::string>(args, "path");
  if (!path || path->empty()) {
    return absl::InvalidArgumentError("Missing mandatory field: path");
  }
  if (absl::StrContains(*path, "..") || (!path->empty() && (*path)[0] == '/')) {
    return absl::PermissionDeniedError("SECURITY_VIOLATION: Path traversal (..) or absolute paths are not allowed.");
  }

  ASSIGN_OR_RETURN(auto start_opt, ParseOptionalInteger(args, "start_line"));
  ASSIGN_OR_RETURN(auto end_opt, ParseOptionalInteger(args, "end_line"));

  std::ifstream in(*path);
  if (!in.is_open()) {
    return absl::NotFoundError(absl::StrCat("Could not open file: ", *path));
  }

  std::vector<std::string> lines;
  std::string line;
  while (std::getline(in, line)) {
    lines.push_back(line);
  }

  const int start_line = start_opt.has_value() ? *start_opt : 1;
  const int end_line = end_opt.has_value() ? *end_opt : static_cast<int>(lines.size());
  if (start_line > end_line) {
    return absl::InvalidArgumentError(
        absl::StrCat("start_line (", start_line, ") cannot be greater than end_line (", end_line, ")"));
  }
  if (start_line > static_cast<int>(lines.size())) {
    return std::string();
  }

  const bool line_numbers = json_get_or<bool>(args, "line_numbers", false);
  const int begin = std::max(1, start_line) - 1;
  const int end_exclusive = std::min(end_line, static_cast<int>(lines.size()));

  std::vector<std::string> out_lines;
  for (int i = begin; i < end_exclusive; ++i) {
    if (line_numbers) {
      out_lines.push_back(absl::StrCat(i + 1, ": ", lines[i]));
    } else {
      out_lines.push_back(lines[i]);
    }
  }

  if (out_lines.empty()) return std::string();
  return absl::StrCat(absl::StrJoin(out_lines, "\n"), "\n");
}

}  // namespace slop
